File: src/2025-03-04-03-28-40--deepseek-chat-0304/go/exercises/practice/ledger/ledger.go
package ledger

import (
	"errors"
	"fmt"
	"sort"
	"strconv"
	"strings"
)

type Entry struct {
	Date        string // "Y-m-d"
	Description string
	Change      int // in cents
}

func sortEntries(entries []Entry) []Entry {
	entriesCopy := make([]Entry, len(entries))
	copy(entriesCopy, entries)
	
	sort.Slice(entriesCopy, func(i, j int) bool {
		if entriesCopy[i].Date != entriesCopy[j].Date {
			return entriesCopy[i].Date < entriesCopy[j].Date
		}
		if entriesCopy[i].Description != entriesCopy[j].Description {
			return entriesCopy[i].Description < entriesCopy[j].Description
		}
		return entriesCopy[i].Change < entriesCopy[j].Change
	})
	
	return entriesCopy
}

func createHeader(locale string) (string, error) {
	switch locale {
	case "nl-NL":
		return fmt.Sprintf("%-10s | %-25s | %s\n", "Datum", "Omschrijving", "Verandering"), nil
	case "en-US":
		return fmt.Sprintf("%-10s | %-25s | %s\n", "Date", "Description", "Change"), nil
	default:
		return "", errors.New("invalid locale")
	}
}

func validateInputs(currency string, locale string) error {
	if currency == "" {
		return errors.New("empty currency")
	}
	if currency != "USD" && currency != "EUR" {
		return errors.New("invalid currency")
	}
	if locale == "" {
		return errors.New("empty locale")
	}
	if locale != "en-US" && locale != "nl-NL" {
		return errors.New("invalid locale")
	}
	return nil
}

func FormatLedger(currency string, locale string, entries []Entry) (string, error) {
	if err := validateInputs(currency, locale); err != nil {
		return "", err
	}
	
	header, err := createHeader(locale)
	if err != nil {
		return "", err
	}
	
	if len(entries) == 0 {
		return header, nil
	}
	
	entriesCopy := sortEntries(entries)
	
	s := header
	// Parallelism, always a great idea
	co := make(chan struct {
		i int
		s string
		e error
	})
	for i, et := range entriesCopy {
		go func(i int, entry Entry) {
			if len(entry.Date) != 10 {
				co <- struct {
					i int
					s string
					e error
				}{e: errors.New("")}
			}
			d1, d2, d3, d4, d5 := entry.Date[0:4], entry.Date[4], entry.Date[5:7], entry.Date[7], entry.Date[8:10]
			if d2 != '-' {
				co <- struct {
					i int
					s string
					e error
				}{e: errors.New("")}
			}
			if d4 != '-' {
				co <- struct {
					i int
					s string
					e error
				}{e: errors.New("")}
			}
			de := entry.Description
			if len(de) > 25 {
				de = de[:22] + "..."
			} else {
				de = de + strings.Repeat(" ", 25-len(de))
			}
			var d string
			if locale == "nl-NL" {
				d = d5 + "-" + d3 + "-" + d1
			} else if locale == "en-US" {
				d = d3 + "/" + d5 + "/" + d1
			}
			negative := false
			cents := entry.Change
			if cents < 0 {
				cents = cents * -1
				negative = true
			}
			var a string
			if locale == "nl-NL" {
				if currency == "EUR" {
					a += "€"
				} else { // USD
					a += "$"
				}
				a += " "
				centsStr := strconv.Itoa(cents)
				switch len(centsStr) {
				case 1:
					centsStr = "00" + centsStr
				case 2:
					centsStr = "0" + centsStr
				}
				rest := centsStr[:len(centsStr)-2]
				var parts []string
				for len(rest) > 3 {
					parts = append(parts, rest[len(rest)-3:])
					rest = rest[:len(rest)-3]
				}
				if len(rest) > 0 {
					parts = append(parts, rest)
				}
				for i := len(parts) - 1; i >= 0; i-- {
					a += parts[i] + "."
				}
				a = a[:len(a)-1]
				a += ","
				a += centsStr[len(centsStr)-2:]
				if negative {
					a += "-"
				} else {
					a += " "
				}
			} else if locale == "en-US" {
				if negative {
					a += "("
				}
				if currency == "EUR" {
					a += "€"
				} else { // USD
					a += "$"
				}
				centsStr := strconv.Itoa(cents)
				switch len(centsStr) {
				case 1:
					centsStr = "00" + centsStr
				case 2:
					centsStr = "0" + centsStr
				}
				rest := centsStr[:len(centsStr)-2]
				var parts []string
				for len(rest) > 3 {
					parts = append(parts, rest[len(rest)-3:])
					rest = rest[:len(rest)-3]
				}
				if len(rest) > 0 {
					parts = append(parts, rest)
				}
				for i := len(parts) - 1; i >= 0; i-- {
					a += parts[i] + ","
				}
				a = a[:len(a)-1]
				a += "."
				a += centsStr[len(centsStr)-2:]
				if negative {
					a += ")"
				} else {
					a += " "
				}
			} else {
				co <- struct {
					i int
					s string
					e error
				}{e: errors.New("")}
			}
			var al int
			for range a {
				al++
			}
			co <- struct {
				i int
				s string
				e error
			}{i: i, s: d + strings.Repeat(" ", 10-len(d)) + " | " + de + " | " +
				strings.Repeat(" ", 13-al) + a + "\n"}
		}(i, et)
	}
	ss := make([]string, len(entriesCopy))
	for range entriesCopy {
		v := <-co
		if v.e != nil {
			return "", v.e
		}
		ss[v.i] = v.s
	}
	for i := 0; i < len(entriesCopy); i++ {
		s += ss[i]
	}
	return s, nil
}

File: src/2025-03-04-03-28-40--deepseek-chat-0304/java/exercises/practice/ledger/src/main/java/Ledger.java
import java.time.LocalDate;
import java.time.format.DateTimeFormatter;
import java.util.ArrayList;
import java.util.List;
import java.util.HashMap;
import java.util.Map;

public class Ledger {
    private static final Map<String, LocaleFormat> LOCALE_FORMATS = new HashMap<>();
    
    static {
        LOCALE_FORMATS.put("en-US-USD", new LocaleFormat("MM/dd/yyyy", "$", ".", ",", "Date       | Description               | Change       "));
        LOCALE_FORMATS.put("nl-NL-USD", new LocaleFormat("dd/MM/yyyy", "$", ",", ".", "Datum      | Omschrijving              | Verandering  "));
        LOCALE_FORMATS.put("en-US-EUR", new LocaleFormat("MM/dd/yyyy", "€", ".", ",", "Date       | Description               | Change       "));
        LOCALE_FORMATS.put("nl-NL-EUR", new LocaleFormat("dd/MM/yyyy", "€", ",", ".", "Datum      | Omschrijving              | Verandering  "));
    }
    public LedgerEntry createLedgerEntry(String date, String description, int change) {
        if (date == null || description == null) {
            throw new IllegalArgumentException("Date and description cannot be null");
        }
        try {
            return new LedgerEntry(LocalDate.parse(date), description, change);
        } catch (Exception e) {
            throw new IllegalArgumentException("Invalid date format", e);
        }
    }

    public String format(String currency, String locale, LedgerEntry[] entries) {
        if (currency == null || locale == null || entries == null) {
            throw new IllegalArgumentException("Arguments cannot be null");
        }
        validateInputs(currency, locale);
        
        LocaleFormat localeFormat = LOCALE_FORMATS.get(locale + "-" + currency);
        StringBuilder result = new StringBuilder(localeFormat.getHeader());

        if (entries.length > 0) {
            List<LedgerEntry> sortedEntries = sortEntries(entries);
            
            for (LedgerEntry entry : sortedEntries) {
                result.append("\n")
                      .append(formatEntry(entry, localeFormat));
            }
        }

        return result.toString();
    }

    private static class LocaleFormat {
        private final String datePattern;
        private final String currencySymbol;
        private final String decimalSeparator;
        private final String thousandSeparator;
        private final String header;

        public LocaleFormat(String datePattern, String currencySymbol, 
                          String decimalSeparator, String thousandSeparator,
                          String header) {
            this.datePattern = datePattern;
            this.currencySymbol = currencySymbol;
            this.decimalSeparator = decimalSeparator;
            this.thousandSeparator = thousandSeparator;
            this.header = header;
        }

        public String getDatePattern() { return datePattern; }
        public String getCurrencySymbol() { return currencySymbol; }
        public String getDecimalSeparator() { return decimalSeparator; }
        public String getThousandSeparator() { return thousandSeparator; }
        public String getHeader() { return header; }
    }

    private void validateInputs(String currency, String locale) {
        if (!currency.equals("USD") && !currency.equals("EUR")) {
            throw new IllegalArgumentException("Invalid currency");
        }
        if (!locale.equals("en-US") && !locale.equals("nl-NL")) {
            throw new IllegalArgumentException("Invalid locale");
        }
    }

    private List<LedgerEntry> sortEntries(LedgerEntry[] entries) {
        List<LedgerEntry> neg = new ArrayList<>();
        List<LedgerEntry> pos = new ArrayList<>();
        
        for (LedgerEntry entry : entries) {
            if (entry.getChange() >= 0) {
                pos.add(entry);
            } else {
                neg.add(entry);
            }
        }

        neg.sort((o1, o2) -> o1.getLocalDate().compareTo(o2.getLocalDate()));
        pos.sort((o1, o2) -> o1.getLocalDate().compareTo(o2.getLocalDate()));

        List<LedgerEntry> all = new ArrayList<>();
        all.addAll(neg);
        all.addAll(pos);
        return all;
    }

    private String formatEntry(LedgerEntry entry, LocaleFormat localeFormat) {
        String date = entry.getLocalDate().format(
            DateTimeFormatter.ofPattern(localeFormat.getDatePattern()));
        
        String description = formatDescription(entry.getDescription());
        String amount = formatAmount(entry.getChange(), localeFormat);
        
        return String.format("%s | %-25s | %13s", date, description, amount);
    }

    private String formatDescription(String description) {
        if (description.length() > 25) {
            return description.substring(0, 22) + "...";
        }
        return description;
    }

    private String formatAmount(double change, LocaleFormat localeFormat) {
        String converted = String.format("%.02f", Math.abs(change / 100));
        String[] parts = converted.split("\\.");
        String amount = formatIntegerPart(parts[0], localeFormat.getThousandSeparator());
        
        String formattedAmount = localeFormat.getCurrencySymbol() + 
            (localeFormat.getDecimalSeparator().equals(",") ? " " : "") +
            amount + localeFormat.getDecimalSeparator() + parts[1];
        
        if (change < 0) {
            formattedAmount = formatNegativeAmount(formattedAmount, localeFormat);
        } else if (localeFormat.getDecimalSeparator().equals(",")) {
            formattedAmount = " " + formattedAmount + " ";
        } else {
            formattedAmount += " ";
        }
        
        return formattedAmount;
    }

    private String formatIntegerPart(String integerPart, String thousandSeparator) {
        StringBuilder amount = new StringBuilder();
        int count = 1;
        for (int ind = integerPart.length() - 1; ind >= 0; ind--) {
            if ((count % 3) == 0 && ind > 0) {
                amount.insert(0, thousandSeparator + integerPart.charAt(ind));
            } else {
                amount.insert(0, integerPart.charAt(ind));
            }
            count++;
        }
        return amount.toString();
    }

    private String formatNegativeAmount(String amount, LocaleFormat localeFormat) {
        if (localeFormat.getDecimalSeparator().equals(".")) {
            return "(" + amount + ")";
        } else {
            return localeFormat.getCurrencySymbol() + " -" + 
                   amount.replace(localeFormat.getCurrencySymbol(), "").trim() + " ";
        }
    }

    public static final class LedgerEntry {
        private final LocalDate localDate;
        private final String description;
        private final double change;

        public LedgerEntry(LocalDate localDate, String description, double change) {
            if (localDate == null || description == null) {
                throw new IllegalArgumentException("Fields cannot be null");
            }
            this.localDate = localDate;
            this.description = description;
            this.change = change;
        }

        public LocalDate getLocalDate() {
            return localDate;
        }

        public String getDescription() {
            return description;
        }

        public double getChange() {
            return change;
        }
    }

}
